docs/architecture/grpc-for-wcf-developers/interface-definition-language.md

```diff
@@ -1,11 +1,31 @@
 ---
 title: Interface Definition Language - gRPC for WCF Developers
-description: TO BE WRITTEN
+description: Introducing the Protocol Buffers IDL.
 author: markrendle
 ms.date: 09/02/2019
 ---
 
 # Interface Definition Language
 
+With WCF, services can expose description metadata using the Web Service Definition Language (WSDL), which is generated dynamically using .NET Reflection at runtime. Developers can use this metadata to generate clients for those services, potentially in other languages if a platform-neutral binding such as SOAP over HTTP is used.
+
+gRPC uses the Interface Definition Language (IDL) from Protocol Buffers. The Protocol Buffers IDL is a custom, platform-neutral language with an open specification. Developers hand-code `.proto` files to describe services along with their inputs and outputs. These `.proto` files can then be used to generate language- or platform-specific stubs for clients and servers, allowing multiple different platforms to communicate. By sharing `.proto` files, teams can generate code to use each others' services without needing to take a code dependency.
+
+One of the advantages of the Protobuf IDL is that as a custom language it enables gRPC to be completely language and platform agnostic, not favoring any technology over another.
+
+The Protobuf IDL is also much easier for humans to both read and write than WSDL. Changing the WSDL of a WCF service typically requires making the changes to the service code itself, running the service and regenerating the WSDL file from the server. By contrast, with a `.proto` file, changes are simple to apply and automatically flow through the generated code. Visual Studio 2019 builds `.proto` files in the background when they are saved; when using other editors such as VS Code the changes will be applied when the project is built.
+
+When compared with XML, messages encoded using Protobuf have many advantages:
+
+- They can be up to 10x smaller.
+- Encoding, decoding and transmitting can be up to 100x faster.
```

Comment:
This also isn't true. When using binary message encoding, WCF's encoding is about 10% slower than protobuf. Also any benchmarks comparing NetTcp with protobuf are showing WCF to be significantly faster.

Reply:
I benchmarked NetTCP against gRPC and gRPC was faster and used less memory. https://visualrecode.com/blog/wcf-vs-grpc-round-2/

Reply:
In your original [blog post](https://unwcf.com/posts/wcf-vs-grpc-round-2/) I listed some problems with your comparison where you were comparing apples to oranges. In your comment reply to me you said you believed neither scenario was using any encryption. WCF definitely was, so presuming gRPC wasn't using any encryption, that will make a massive difference. If gRPC is using encryption, it's using TLS whereas you are using Windows auth for WCF which translates to using NegotiateStream. This are different encryption mechanisms so there could be a difference in performance there. Also, you were running WCF using workstation GC and gRPC using server GC. This again can make a significant difference, especially for server scenarios. Workstation GC has a single heap that all processors have to contend with to get access to. It also means when you do need to do a GC pause, you have a single core doing the work instead of all your cores so pauses are longer.  
And if all those problems aren't enough, your own results with a hamstrung WCF only show grpc to be 1.11X faster, not 100X.

Reply:
Look, I'm not an expert at configuring WCF to get the best performance, but to be fair, I'm not an expert at configuring gRPC to get the best performance either. They were both running on whatever their default settings were, which for gRPC is obviously HTTP/2 over TLS, I don't know what it is for NetTCP.

_Anyway_, it's irrelevant here because the comparison is not gRPC versus WCF-with-NetTCP-binding. It's Protobuf versus XML.

> **When compared with XML**, messages encoded using Protobuf have many advantages:
> 
> - They can be up to 10x smaller.
> - Encoding, decoding and transmitting can be up to 100x faster.